Data Alignment
=============

**Goal**

This stage of the project involved the aligning of the previously processed modern data (Australian and American) and Ancient data against a created database. The databases as created through Malt Build each contaim HOMD and the NCBI accession database as well as one or a combination of viral databases (listed below). By aligning the sample sequences against these databases it is possible to determine what species of virus are present in the human oral virome at varying time points in varying geographical locations. Through this reference mapping it is possible to confirm viral identification is possible within dental caclulus as well as to determine the diversity of the virome community. 

**Tasks**

1. [Collect Database Elements](https://github.com/darcey-watson/Viral-Content-Project/blob/master/2.%20Data%20Alignment.md#1-collecting-database-elements)
2. [Construct Databases](https://github.com/darcey-watson/Viral-Content-Project/blob/master/2.%20Data%20Alignment.md#2constructing-the-databases)
3. [Database Alignment](https://github.com/darcey-watson/Viral-Content-Project/blob/master/2.%20Data%20Alignment.md#3database-alignment)
4. [Aligned Database Conversion](https://github.com/darcey-watson/Viral-Content-Project/blob/master/2.%20Data%20Alignment.md#4-aligned-database-conversion)

## 1. Collecting Database Elements
#### The following databases were collected to construct the databases:
 - HOMD (Human Oral Microbiome Database)
	- Obtained from: 
	- Contains 770 bacterial species
	- included to prevent false positive alignments caused by mis-mapping bacterial reads to viral reads.
 - NCBI nucleotide accessions 
 	- Obtained from: http://ab.inf.uni-tuebingen.de/data/software/megan6/download/welcome.html 
	- Contains all annotated accession sequences for viruses currently identified in NCBI
 - NCBIcomplete and NCBIincomplete (viral databases)
	 	 - Direct NCBI download of all viral genomes
	* To obtain these genomes download the following program:
		
      		conda install -c bioconda ncbi-genome-download
   * To then commence the download of all of the viral genomes complete the following:
   For NCBIcomplete, enter the following:
		
    		ncbi-genome-download --format fast viral --assembly-level complete
    
    and for NCBIincomplete, enter the following 
    
       		ncbi-genome-download --format fast viral 
    
 - Virusite
	 - Obtained from: http://www.virusite.org/index.php?nav=download
	 - Version: 2019.1
	 - Composed of 9233 viral reads from several online viral repositories including: UniProtKB and ViralZone
	 - Focuses on viral genes, proteins and genomes from viruses, viroids and satellites
- RVDB
	- Obtained from: https://hive.biochemistry.gwu.edu/rvdb
	- Version: 16.0
	- Consists of 717,145 eukaryotic viral, virus-related and virus-like sequences, regardless of size.


## 2. Constructing the Databases
To create the database that the modern data needs to be compared to we will use MALT through its MALT build function. 
This will be completed using the following commands:

#### MALT was installed via:

#### To run the reads the following commands were entered:

    module load malt

#### The script to be entered to create the initial database (HOMD and NCBI only) is the following:
    #!/bin/bash
    #SBATCH -p highmem
    #SBATCH -N 1
    #SBATCH -n 30
    #SBATCH --time=05:00:00
    #SBATCH --mem=250GB
    
    #Notification configuration
    #SBATCH --mail-type=END
    #SBATCH --mail-type=FAIL
    #SBATCH --mail-user=(email address)
 
    malt-build -i (input database: HOMD) \
    -d (output) \
    -a2taxonomy (annotated accession numbers: NCBI) \
    --sequenceType DNA -t 30 -v

#### The script to be entered to create the viral database (HOMD, NCBI and Virusite) is the following:

    #!/bin/bash
    #SBATCH -p highmem
    #SBATCH -N 1
    #SBATCH -n 30
    #SBATCH --time=05:00:00
    #SBATCH --mem=250GB
    
    #Notification configuration
    #SBATCH --mail-type=END
    #SBATCH --mail-type=FAIL
    #SBATCH --mail-user=(email address)
 
    malt-build -i (input: HOMD and Virusite) \
    -d (output) \
    -a2taxonomy (annotated accession numbers: NCBI) \
    --sequenceType DNA -t 30 -v 
    
* To complete this step it is first neccessary to move all of the input sources i.e. Homd and Virusite into the same directory. Once inside they must be the same file type so as to ensure both files are recognised. In this case the HOMD database was renamed .fasta to make it recognisable. This step will allow for only 1 input to be entered with all data obtained, thus making the process much simpler.

#### Program Version:

## 3. Database Alignment
To align the modern data with the above created aligned database, MALT will again be used with its MALTrun funciton. This will result in ()(files that have compared and aligned the modern data to the developed database so that each of the sequences in the modern data samples can be aligned to its closest annotated sequences from the database

#### Malt has already been installed (see above step 2.)

#### To run the reads the following commands need to be entered:

    module load malt
    sbatch (Jobscript name)

#### The script to be entered to run the alignment is the following:

    #!/bin/bash
    #SBATCH -p highmem
    #SBATCH -N 1
    #SBATCH -n 32
    #SBATCH --time=30:00:00
    #SBATCH --mem=1500GB
    
    #Notification configuration
    #SBATCH --mail-type=END
    #SBATCH --mail-type=FAIL
    #SBATCH --mail-user=(email address)
    FG
    malt-run \
    -i (input: data) \
    -a (output) \
    -ou (output: same as above) \
    -at SemiGlobal -f Text -d (database: output from MALT Build) \
    -mem load --mode BlastN -t 32 -v
    ~
    ~
    ~


## 4. Aligned Database Conversion

#### Blast2RMA was already installed as part of the ()()( package. It is possible to check whether Blast2RMA is already installed via:

    module spider blast2rma

#### To run the reads the following commands need to be entered:

    module load malt
    module load parallel

#### The script to be entered to run the alignment is the following:

    #!/bin/bash
    #SBATCH -p highmem
    #SBATCH -N 1
    #SBATCH -n 32
    #SBATCH --time=15:00:00
    #SBATCH --mem=500GB
    
    #Notification configuration
    #SBATCH --mail-type=END
    #SBATCH --mail-type=FAIL
    #SBATCH --mail-user=(email address)
    
    for i in (input) \
    do b=${i%.fa*}.blastn.gz; echo blast2rma -r $i -i $b \
    -o (output) \
    -ms 44 -me 0.01 -f BlastText -a;g weighted -lcp 80; done | parallel -j28 --delay 6

* The input for blast2RMA must consist of 

#### Program Version:
