## Individual Viral Genome Alignments and Classification

#### Goal
This stage of the project involves the completion of individual viral genome alignments to allow for further identification and classification of viral species within the datasets.

#### Tasks
1. [Identify most prominent species within each dataset from MEGAN6 output](https://github.com/darcey-watson/Viral-Content-Project/blob/master/4.%20Individual%20Viral%20genome%20alignments%20and%20classification.md#1-identify-prominent-species)
2. [Download specific viral references of those viral species](https://github.com/darcey-watson/Viral-Content-Project/blob/master/4.%20Individual%20Viral%20Genome%20Alignments%20and%20Classification.md#2-download-specific-viral-references)
3. [Run all dataset sequences and viral references through BWA](https://github.com/darcey-watson/Viral-Content-Project/blob/master/4.%20Individual%20Viral%20Genome%20Alignments%20and%20Classification.md#3-bwa-alignment) 
4. [Analyse output through Geneious](https://github.com/darcey-watson/Viral-Content-Project/blob/master/4.%20Individual%20Viral%20Genome%20Alignments%20and%20Classification.md#4-geneious-analysis)us

### 1. Identify prominent species 
This step requires analysis of MEGAN6 viral sequence results. Through investigating the viral reads specified for each of the datasets it is possible to observe the viral species with the most assigned reads. By opening up individual samples and selecting those species which indicated the highest proportion of assigned reads it is possible to 'Inspect Alignment' and 'Show Alignment'. To finally select the most prominent species within each dataset several criteria should be met. Firstly, a species should only be investigated if it has over 100 assigned reads (this is dependent on total overall reads) or has the majority of reads assigned within any dataset. Next, the 'identities' within the 'Inspect Alignment' option should be above 85 % match with less than 5 % gaps. Finally, the alignment seen through 'Show Alignment' should be evenly distributed with minimal clustering over possible gene sites. Once these viruses have been selected it is possible to further access them

### 2. Download specific viral references
To download the exact viral reference sequences matched during the MEGAN6 alignment go to the 'Inspect Alignment tab and the accession number for either GENBANK or , will be listed. See image below to identify accession number location. Copy this accession number and search through the GENBANK database on NCBI. Once the appropriate page has download select to download the sequence in a fasta file format. Once all neccessary reference sequences are downloaded combine them in a central viral reference folder to act as the reference input to the BWA command

### 3. BWA alignment 
To first download BWA enter the following command:

    conda create -n BWA -c bioconda bwa
In order to run the BWA script an additional script located in a text file is needed. This text file needs to contain the merge contigs script which will then be given an absolute path to the BWA script. The merge contigs script is below:

```bash
function print_n(n)
 {
    text = ""
    for(j=0;j<n;j++) text=text"N"
    return text
 }

BEGIN {n=1} 
$0 ~ "^>" { 
    if(n<=i) {
        header = header""substr($0,2)" ";
        if(n > 1) seq=seq""print_n(1000)
        n++;
    } 
    else {
        print ">"header"\n"seq; 
        n=1; 
        header=substr($0,2); 
        seq=""
    } 

    next;
} 
{seq=seq$0} 
END {print ">"header"\n"seq;}
```
To complete the processing of BWA the following script needs to be entered:

    ### Competitive mapping bash script, Raphael Eisenhofer 2019-August ###
    ## Reference files need to be in fna.gz format
    ## You'll need to download mergeContigs.awk and set an absolute path to it
    
    echo " _______  __       _______. _______ .__   __.  __    __    ______    _______  _______ .______         ";
    echo "|   ____||  |     /       ||   ____||  \ |  | |  |  |  |  /  __  \  |   ____||   ____||   _  \        ";
    echo "|  |__   |  |    |   (----\`|  |__   |   \|  | |  |__|  | |  |  |  | |  |__   |  |__   |  |_)  |       ";
    echo "|   __|  |  |     \   \    |   __|  |  . \`  | |   __   | |  |  |  | |   __|  |   __|  |      /        ";
    echo "|  |____ |  | .----)   |   |  |____ |  |\   | |  |  |  | |  \`--'  | |  |     |  |____ |  |\  \----.   ";
    echo "|_______||__| |_______/    |_______||__| \__| |__|  |__|  \______/  |__|     |_______|| _| \`._____|   ";
    echo "                                                                                                      ";
    echo " __  .__   __.  _______   __    __       _______..___________..______       __   _______      _______.";
    echo "|  | |  \ |  | |       \ |  |  |  |     /       ||           ||   _  \     |  | |   ____|    /       |";
    echo "|  | |   \|  | |  .--.  ||  |  |  |    |   (----\`\`---|  |----\`|  |_)  |    |  | |  |__      |   (----\`";
    echo "|  | |  . \`  | |  |  |  ||  |  |  |     \   \        |  |     |      /     |  | |   __|      \   \    ";
    echo "|  | |  |\   | |  '--'  ||  \`--'  | .----)   |       |  |     |  |\  \----.|  | |  |____ .----)   |   ";
    echo "|__| |__| \__| |_______/  \______/  |_______/        |__|     | _| \`._____||__| |_______||_______/    ";
    echo "                                                                                                      ";
    
    #Set up user defined variables:
    
    	echo "Please specify path to reference sequences for competitive mapping."
    		read RefDir
    
    	echo "Please specify the directory containing the reads that you wish to map."
    		read ReadDir
    		
    	echo "Please specify the output directory."
    		read OutDir
    
    	echo "Please specify the minimum mapping quality that you wish to filter the BAM files with."
    		read MQ
    		
    	echo "Please specify the number of threads you'd like to use."
    		read threads
    
    ##Need to massage assemblies that contain scaffolds/assemblies, as when we split the concatenated reference by assembly, we don't want 10's to 100's of separate BAM files... Also makes it easier to visualise/calculate coverage/depth with tools like qualimap.
    #For each assembly, merge contigs (with 500 bp of N's between each to stop mapping issues) and headers. This is a akin to creating a pseudo-complete genome. Note mergeConrigs.awk was written by finswimmer (https://www.biostars.org/p/379245/)
    	echo "Massaging non-complete assemblies..."
    	for i in $RefDir/*.gz; do zcat $i | awk -v i=200 -f (absolute path to mergeContigs.awk) > ${i/.fna.gz/-fixed.fna}; done
    			
    #The assembly headers need to be cleaned up for use in ANVI'O, which doesn't like funky characters. This step cleans them up, making them usable by ANVI'O in downstream steps:
    	for i in $RefDir/*.fna; do sed -i'' 's/\..*//g' $i; done
    	#for i in $RefDir/*.fna; do sed 's/,/_/g' $i | sed 's/\./_/g' | sed 's/ /_/g' | sed 's/:/_/g' | sed 's/;/_/g' > ${i/.fna/-header.fna}; done
    	
    	pigz $RefDir/*.fna
    		
    #Concatenate references and index using BWA:
    
    	echo "Concatenating references..."
    	cat $RefDir/*-fixed.fna.gz > $RefDir/CattedRefs.fna.gz
    	
    	echo "Indexing concatenated references..."
    	bwa index $RefDir/CattedRefs.fna.gz
    	
    #Map reads to BWA index, convert to SAM, then BAM, keeping headers and only mapped reads and sorting:
    
    	echo "Mapping reads..."
    	mkdir $OutDir
    	mkdir $OutDir/SAIs
    	mkdir $OutDir/Deduped-RAW-BAMs
    	mkdir $OutDir/dedup-log
    	mkdir $OutDir/Deduped-MQ$MQ-Split-BAMs
    	
    	for i in $ReadDir/*.fastq.gz; \
    	do bwa aln -l 1024 -t $threads $RefDir/CattedRefs.fna.gz $i > $OutDir/SAIs/$(basename ${i/.fastq.gz/.sai}); done
    	
    	echo "Converting sai to SAM/BAM, keeping headers and sorting the BAM file..."
    	for i in $OutDir/SAIs/*.sai; \
    	do bwa samse $RefDir/CattedRefs.fna.gz $i $ReadDir/$(basename ${i/.sai/.fastq.gz}) \
    	 | samtools view -S -h -b -F 4 -q $MQ - \
    	 | samtools sort - $OutDir/Deduped-RAW-BAMs/$(basename ${i/.sai/}); done
    	
    	echo "Removing duplicates..."
    	for i in $OutDir/Deduped-RAW-BAMs/*.bam; \
    	do dedup -i $i -m -o $OutDir/dedup-log/; done
    	
    	#Remove non-deduped BAMs
    	rm $OutDir/Deduped-RAW-BAMs/*.bam
    	
    	#Sort BAMs
    	for i in $OutDir/dedup-log/*rmdup.bam; \
    	do samtools sort $i $OutDir/Deduped-RAW-BAMs/$(basename ${i/_rmdup.bam/-sorted}); done
    	
    	#Index BAMs
    	for i in $OutDir/Deduped-RAW-BAMs/*.bam; \
    	do samtools index $i; done
    	
    	#Remove unsorted BAMs
    	rm $OutDir/dedup-log/*rmdup.bam
    	
    	
    #Split BAMs by reference and run qualimap on each one
    
    	echo "Splitting BAMs by reference..."
    	#Get genome names and place them in a file to be read by next step:
    	for i in $OutDir/Deduped-RAW-BAMs/*.bam; do samtools view -H $i \
    	 | grep @SQ \
    	 | awk '{print $2}' \
    	 | sed 's/SN://g' > $i-RefHeaders.txt; done
    	#Split SAM files by genome names in RefHeaders.txt (need to split SAM, as BAM has indexing which screws up when you manipulate the headers):
    	for BAM in $OutDir/Deduped-RAW-BAMs/*.bam; \
    	do while read REF; \
    	do samtools view -h $BAM $REF > ${BAM/.bam/-$REF.sam}; done < $BAM-RefHeaders.txt; done
    	
    	#Parse through split SAM files, removing the genome name SAM header entries not in the file, outputting BAM files:
    	for REFHEADER in $OutDir/Deduped-RAW-BAMs/*RefHeaders.txt; \
    	do while read REF; \
    	do for SAM in ${REFHEADER/.bam-RefHeaders.txt/}-$REF.sam; \
    	do samtools view -S -h $SAM \
    	 | awk -v ref="$REF" 'BEGIN { FS = OFS = "\t"; } {if ($1 == "@SQ" && $2 != "'SN":"$REF'") { } else print; }' \
    	 | samtools view -h -S -b - > $OutDir/Deduped-MQ$MQ-Split-BAMs/$(basename ${SAM/.sam/.bam}); done; done < $REFHEADER; done
    	
    	#Clean up unneeded SAM files/test files:
    	rm $OutDir/Deduped-RAW-BAMs/*.sam
    	rm $OutDir/Deduped-RAW-BAMs/*RefHeaders.txt
    
    
    		
    	#Remove catted reference file...
    	rm $RefDir/CattedRefs.fna.gz*
    
    		
    #Say goodbye
    	echo 	   "##"
    	echo      "####"
    	echo     "######"
    	echo    "########"
    	echo   "##########"
    	echo  "############"
    	echo "##############"
    	echo "WORK COMPLETE!"

Following the completion of the BWA script (which is able to run on local machines) you will recieve an output of split BAM files which are then able to be entered into Geneious.

### 4. Geneious analysis
To analyse the BAM files through Geneious it is necessary to load both the original viral references and split BAM outputs into the GUI. Once loaded select the 'align to' option and furthermore the 'align to reference' option, several statistics of the sequences will be shown. 

Note: Geneious does require a subscription 
