# Viral-Content-Project

Identifying viral content and diversity in ancient and modern dental calculus samples

**Summary of Project**

Investigating the viral content of both modern and ancient dental calculus samples collected from varying locations across the globe. Observing the diversity and complexity of the human oral virome and how these conditions have changed over time. 

**Project Context**

Track changes in viral population diversity and viral genomes dependent on time and geography 

**Project Aims**

 1. To detect and identify viruses within modern day dental calculus samples
 2. To determine the diversity of viruses within dental calculus and compare the extent of diversity observed between Australian and European samples
 3. To determine if viruses present in modern day Australian samples can be detected in ancient European individuals
 4. To analyse whether the ancient oral virome changes based on time and geography

**Approach Summary**
Bioinformatic analysis techniques will be used to address the above listed aims. These will include the processing and analysis of already obtained and sequenced data (from ACAD). This processing will involve two main mechanisms. 

**Data**
 - ACAD Sequence Files
	 - Received ()() sequences from modern individuals, known as ModernACAD_oral data????
 - HMP Sequence Files
	 - sequence data obtained from the Human Microbiome Project, including data obtained from the three oral sites: Tongue-, and .
	 - there were a total of ()()() samples. a total of 5 were selected to continue with, these include:()(). from these 5 samples a total of 20 million reads were selected from each using the following script::::()()()(()( 
 - Metadata from ACAD
	 - Metadata and sample information obtained from the ACAD sequence files collection process. This information will allow for socio-economic and other variable influences over the data to be understood and interpreted in context to the data

**Comparative Data**
 - HOMD Data
	 - To compare the ACAD Sequence files and HMP sequence files I have used the human oral microbiome database which i combined with:
 - Ncbi accession 
	 - this file has annotated accession numbers from ncbi which are crossed with the HOMD data and allows for identification of any corresponding sequences within the experimental data
 - Virusite 
	 - This online database is comprised of combined databases from ()()()()(()

**Acknowledgements**

My supervisors: Dr Laura Weyrich and Dr Raphael Eisenhofer Philipona 

The team at the Australian Centre for Ancient DNA

The University of Adelaide Honours coordinators: Dr Renato Morona, Dr Rob Richards and Dr Murray Whitelaw. 
